internal/controller/rabbitmq/rabbitmqvhost_controller.go

```diff
@@ -0,0 +1,233 @@
+		return ctrl.Result{}, err
+	}
+
+	// Get admin credentials
+	rabbitSecret, _, err := oko_secret.GetSecret(ctx, h, rabbit.Status.DefaultUser.SecretReference.Name, instance.Namespace)
+	if err != nil {
+		instance.Status.Conditions.Set(condition.FalseCondition(rabbitmqv1.VhostReadyCondition, condition.ErrorReason, condition.SeverityWarning, rabbitmqv1.VhostReadyErrorMessage, err.Error()))
+		return ctrl.Result{}, err
+	}
+
+	// Create API client
+	tlsEnabled := rabbit.Spec.TLS.SecretName != ""
+	protocol := "http"
+	managementPort := "15672"
+	if tlsEnabled {
+		protocol = "https"
+		managementPort = "15671"
+	}
+	baseURL := fmt.Sprintf("%s://%s:%s", protocol, string(rabbitSecret.Data["host"]), managementPort)
+	apiClient := rabbitmqapi.NewClient(baseURL, string(rabbitSecret.Data["username"]), string(rabbitSecret.Data["password"]), tlsEnabled)
+
+	// Create vhost
+	vhostName := instance.Spec.Name
+	if vhostName == "" {
+		vhostName = "/"
+	}
+
+	if vhostName != "/" {
+		err = apiClient.CreateOrUpdateVhost(vhostName)
+		if err != nil {
+			instance.Status.Conditions.Set(condition.FalseCondition(rabbitmqv1.VhostReadyCondition, condition.ErrorReason, condition.SeverityWarning, rabbitmqv1.VhostReadyErrorMessage, err.Error()))
+			return ctrl.Result{}, err
+		}
+	}
+
+	instance.Status.Conditions.MarkTrue(rabbitmqv1.VhostReadyCondition, rabbitmqv1.VhostReadyMessage)
+	instance.Status.Conditions.MarkTrue(condition.ReadyCondition, condition.ReadyMessage)
+
+	return ctrl.Result{}, nil
+}
+
+func (r *RabbitMQVhostReconciler) reconcileDelete(ctx context.Context, instance *rabbitmqv1.RabbitMQVhost, h *helper.Helper) (ctrl.Result, error) {
+	// If protection finalizer exists, check if still actively used
+	if controllerutil.ContainsFinalizer(instance, rabbitmqv1.VhostFinalizer) {
+		// Check if still actively used by TransportURL
+		for _, owner := range instance.OwnerReferences {
+			if owner.Controller != nil && *owner.Controller && owner.Kind == "TransportURL" {
+				transportURL := &rabbitmqv1.TransportURL{}
+				if err := r.Get(ctx, types.NamespacedName{Name: owner.Name, Namespace: instance.Namespace}, transportURL); err == nil && transportURL.DeletionTimestamp.IsZero() {
+					vhost := transportURL.Spec.Vhost
+					if vhost == "" {
+						vhost = "/"
+					}
+					expectedRef := ""
+					if vhost != "/" {
+						expectedRef = fmt.Sprintf("%s-%s-vhost", transportURL.Name, vhost)
+					}
+					if expectedRef == instance.Name {
+						return ctrl.Result{RequeueAfter: time.Duration(2) * time.Second}, nil
+					}
+				}
+			}
+		}
+		// Check if any active users reference this vhost
+		userList := &rabbitmqv1.RabbitMQUserList{}
+		if err := r.List(ctx, userList, client.InNamespace(instance.Namespace)); err == nil {
+			for _, user := range userList.Items {
+				if user.Spec.VhostRef == instance.Name && user.DeletionTimestamp.IsZero() {
```

Comment:
in theory the user should probably have a finalizer on the vhost it is on and remove it in its delete process when it got deleted. its just a lot of finalizers.